### PR TITLE
Committing changes related to gRPC to Rest migration for create progr…

### PR DESCRIPTION
**Pull request by Azher2Ali**
Committing changes related to gRPC to Rest migration for create program api

Adding screenshot for response tested from local
Screenshot 2023-07-18 at 12 58 37 PM

### CODE REVIEW

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+@Slf4j
+@RestController
+@RequestMapping("/program")
+public class ProgramController {
+
+  @Autowired private ProgramServiceFacade serviceFacade;
+  @Autowired private Grpc2JsonConverter grpc2JsonConverter;
+  @Autowired private AuthorizationService authorizationService;
+
+  @PostMapping(value = "/createProgram")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:32]: Most endpoints for an entity don't need a custom path, the HTTP request verb should indicate the action.

Create should be a POST request to `/programs`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@Slf4j
+@RestController
+@RequestMapping("/program")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:25]: Following REST naming conventions lets make this plural: `/programs`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    CreateProgramRequest request =
+        grpc2JsonConverter.fromJson(
+            grpc2JsonConverter.getJsonFromObject(createProgramRequestDTO),
+            CreateProgramRequest.class);
+    CreateProgramResponse response = serviceFacade.createProgram(request);
+    return new ResponseEntity(
+        grpc2JsonConverter.prepareCreateProgramResponse(response), HttpStatus.CREATED);
+  }
+
+  @DeleteMapping(value = "/removeProgram")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:45]: DELETE request to `/programs/{programShortName}`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+              grpc2JsonConverter.getJsonFromObject(removeProgramRequestDTO),
+              RemoveProgramRequest.class);
+      serviceFacade.removeProgram(request);
+    } catch (EmptyResultDataAccessException | InvalidDataAccessApiUsageException | IOException e) {
+      log.error("Exception throw in removeProgram: {}", e.getMessage());
+      throw new NotFoundException(ExceptionUtils.getStackTrace(e));
+    }
+  }
+
+  @PutMapping(value = "/updateProgram")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:61]: PUT request to `/programs`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      response = serviceFacade.updateProgram(request);
+    } catch (NotFoundException | NoSuchElementException | IOException e) {
+      log.error("Exception throw in updateProgram: {}", e.getMessage());
+      throw new NotFoundException(ExceptionUtils.getStackTrace(e));
+    }
+    return new ResponseEntity(
+        grpc2JsonConverter.prepareUpdateProgramResponse(response), HttpStatus.OK);
+  }
+
+  @GetMapping(value = "/getProgram/{shortName}")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:81]: We can remove the `/getProgram` path so its just GET on `/programs/{shortName}`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      @PathVariable(value = "shortName", required = true) String shortName) throws IOException {
+    authorizationService.requireProgramUser(shortName);
+    GetProgramRequest request =
+        GetProgramRequest.newBuilder().setShortName(StringValue.of(shortName)).build();
+    GetProgramResponse response = serviceFacade.getProgram(request);
+    return new ResponseEntity(
+        grpc2JsonConverter.prepareGetProgramResponse(response), HttpStatus.OK);
+  }
+
+  @PostMapping(value = "/activateProgram")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:92]: This one does need an additional path, can we just use `/activate`?

POST `/programs/activate`

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      response = serviceFacade.activateProgram(request);
+    } catch (NotFoundException | NoSuchElementException | IOException e) {
+      log.error("Exception throw in activateProgram: {}", e.getMessage());
+      throw new NotFoundException(ExceptionUtils.getStackTrace(e));
+    }
+    return new ResponseEntity(
+        grpc2JsonConverter.prepareGetProgramResponse(response), HttpStatus.OK);
+  }
+
+  @GetMapping(value = "/listProgram")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/controller/ProgramController.java:112]: This should be the behaviour of GET `/programs` with no `{shortName}` path parameter provided

**Review by joneubank — CHANGES_REQUESTED** [review]
Some changes for the request paths to try to follow REST API naming conventions

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 1 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Removed /createProgram from the request mapping

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 2 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Modified Request Mapping to /programs

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 3 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Modified request mapping of Delete API to /programs/{programShortName}

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 4 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Remove /updateProgram request mapping

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 5 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: removed /getProgram mapping and added {shortName} as path variable

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 6 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Modified request mapping to /activate

File: src/main/java/org/icgc/argo/program_service/controller/ProgramController.java (hunk 7 above)
Comment (Azher2Ali) [reply src/main/java/org/icgc/argo/program_service/controller/ProgramController.java]: Modified list api request mapping to /programs

**Review by joneubank — APPROVED** [review]

**Pull request merged**